File: archive/p/pascal/maximum_array_rotation.pas
program MaximumArrayRotation;

{$mode objfpc}{$H+}

uses
   Classes,
   Generics.Collections,
   SysUtils;

procedure ShowUsage;
begin
   Writeln('Usage: please provide a list of integers (e.g. "8, 3, 1, 2")');
   Halt(1);
end;

type
   TIntegerList = specialize TList<integer>;

function ParseIntegerList(const S: string): TIntegerList;
var
   Tokens: TStringArray;
   Token: string;
   Value: integer;
begin
   if S.Trim = '' then
      ShowUsage;

   Tokens := S.Split([',']);
   if Length(Tokens) = 0 then
      ShowUsage;

   Result := TIntegerList.Create;
   for Token in Tokens do
   begin
      if not TryStrToInt(Trim(Token), Value) then
      begin
         Result.Free;
         ShowUsage;
      end;
      Result.Add(Value);
   end;
end;

function MaximumRotationSum(const Numbers: TIntegerList): integer;
var
   N, i, TotalSum, CurrentWeightedSum, MaxWeightedSum: integer;
begin
   N := Numbers.Count;
   if N = 0 then
      ShowUsage;

   TotalSum := 0;
   CurrentWeightedSum := 0;
   for i := 0 to N - 1 do
   begin
      TotalSum += Numbers[i];
      CurrentWeightedSum += Numbers[i] * i;
   end;

   MaxWeightedSum := CurrentWeightedSum;

   for i := 1 to N - 1 do
   begin
      CurrentWeightedSum := CurrentWeightedSum + TotalSum - N * Numbers[N - i];
      if MaxWeightedSum < CurrentWeightedSum then
         MaxWeightedSum := CurrentWeightedSum;
   end;

   Result := MaxWeightedSum;
end;

var
   InputList: TIntegerList;
begin
   if ParamCount <> 1 then
      ShowUsage;

   InputList := ParseIntegerList(ParamStr(1));
   try
      Writeln(MaximumRotationSum(InputList));
   finally
      InputList.Free;
   end;
end.

File: archive/v/visual-basic/maximum-array-rotation.vb
Public Module MaximumArrayRotation

    Private Sub ShowUsage()
        Console.Error.WriteLine(
            "Usage: please provide a list of integers (e.g. ""8, 3, 1, 2"")"
        )
        Environment.Exit(1)
    End Sub


    Private Function ParseList(input As String) As List(Of Integer)

        If String.IsNullOrWhiteSpace(input) Then
            ShowUsage()
        End If

        Dim result As New List(Of Integer)

        For Each part In input.Split(","c, StringSplitOptions.RemoveEmptyEntries Or StringSplitOptions.TrimEntries)

            Dim value As Integer

            If Not Integer.TryParse(part, value) OrElse value < 0 Then
                ShowUsage()
            End If

            result.Add(value)

        Next

        If result.Count = 0 Then
            ShowUsage()
        End If

        Return result

    End Function


    Private Function MaximumRotationSum(numbers As List(Of Integer)) As Integer

        Dim n As Integer = numbers.Count

        If n = 0 Then
            ShowUsage()
        End If

        Dim totalSum As Integer = 0
        Dim currentWeightedSum As Integer = 0

        For i As Integer = 0 To n - 1
            totalSum += numbers(i)
            currentWeightedSum += numbers(i) * i
        Next

        Dim maxWeightedSum As Integer = currentWeightedSum

        For i As Integer = 1 To n - 1

            currentWeightedSum =
                currentWeightedSum + totalSum - n * numbers(n - i)

            If currentWeightedSum > maxWeightedSum Then
                maxWeightedSum = currentWeightedSum
            End If

        Next

        Return maxWeightedSum

    End Function


    Public Function Main(args As String()) As Integer

        If args.Length <> 1 Then
            ShowUsage()
        End If

        Dim inputList = ParseList(args(0))

        Console.WriteLine(MaximumRotationSum(inputList))

        Return 0

    End Function

End Module